modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/ByteBufferWalIterator.java

```diff
@@ -0,0 +1,169 @@
+ */
+
+package org.apache.ignite.internal.processors.cache.persistence.wal;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.pagemem.wal.record.WALRecord;
+import org.apache.ignite.internal.processors.cache.GridCacheSharedContext;
+import org.apache.ignite.internal.processors.cache.persistence.wal.record.HeaderRecord;
+import org.apache.ignite.internal.processors.cache.persistence.wal.serializer.RecordSerializer;
+import org.apache.ignite.internal.processors.cache.persistence.wal.serializer.RecordSerializerFactory;
+import org.apache.ignite.internal.processors.cache.persistence.wal.serializer.RecordSerializerFactoryImpl;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteBiPredicate;
+import org.apache.ignite.lang.IgniteBiTuple;
+import org.jetbrains.annotations.NotNull;
+
+import static org.apache.ignite.internal.pagemem.wal.record.WALRecord.RecordType.HEADER_RECORD;
+import static org.apache.ignite.internal.processors.cache.persistence.wal.serializer.RecordV1Serializer.HEADER_RECORD_SIZE;
+
+/** Byte Buffer WAL Iterator */
+public class ByteBufferWalIterator extends AbstractWalRecordsIteratorAdapter {
+    /** */
+    private static final long serialVersionUID = 0L;
+
+    /** */
+    private final ByteBuffer buf;
+
+    /** */
+    private final RecordSerializer serializer;
+
+    /** */
+    private final ByteBufferBackedDataInputImpl dataInput;
+
+    /** */
+    private boolean hdrChecked;
+
+    /** */
+    public ByteBufferWalIterator(
+        @NotNull IgniteLogger log,
+        @NotNull GridCacheSharedContext<?, ?> cctx,
+        @NotNull ByteBuffer byteBuf,
+        int ver) throws IgniteCheckedException {
+        this(log, cctx, byteBuf, ver, null);
+    }
+
+    /** */
+    public ByteBufferWalIterator(
+        @NotNull IgniteLogger log,
+        @NotNull GridCacheSharedContext<?, ?> cctx,
+        @NotNull ByteBuffer byteBuf,
+        int ver,
+        IgniteBiPredicate<WALRecord.RecordType, WALPointer> readTypeFilter)
+        throws IgniteCheckedException {
+        super(log);
+
+        buf = byteBuf;
+
+        RecordSerializerFactory rsf = new RecordSerializerFactoryImpl(cctx, readTypeFilter);
+
+        serializer = rsf.createSerializer(ver);
+
+        dataInput = new ByteBufferBackedDataInputImpl();
+
+        dataInput.buffer(buf);
+
+        advance();
+    }
+
+    /** */
+    private IgniteBiTuple<WALPointer, WALRecord> advanceRecord() throws IgniteCheckedException {
+        if (!buf.hasRemaining())
+            return null;
+
+        IgniteBiTuple<WALPointer, WALRecord> result;
+
+        try {
+            if (!hdrChecked) {
+                IgniteBiTuple<WALPointer, WALRecord> hdr = tryToReadHeader();
+
+                if (hdr != null)
+                    return hdr;
```

Comment:
Missed check by `recordTypeFilter` if it's specified. User will receive unexpected record then.